i6_models/parts/conformer/mhsa_rel_pos.py

```diff
@@ -0,0 +1,248 @@
+        att_weights_dropout: attention weights dropout
+        learnable_pos_emb: whether to use learnable relative positional embeddings instead of fixed sinusoidal ones
+        rel_pos_clip: maximal relative postion for embedding
+        with_pos_bias: whether to add additional position bias terms to the attention scores
+        pos_emb_dropout: dropout for the positional embeddings
+        dropout: multi-headed self attention output dropout
+        dropout_broadcast_axes: string of axes to which dropout is broadcast, e.g. "T" for broadcasting to the time axis
+                                setting to None to disable broadcasting
+    """
+
+    input_dim: int
+    num_att_heads: int
+    att_weights_dropout: float
+    dropout: float
+    learnable_pos_emb: bool = True
+    rel_pos_clip: Optional[int] = None
+    with_pos_bias: bool = False
+    pos_emb_dropout: float = 0.0
+    dropout_broadcast_axes: Optional[str] = None
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        assert self.input_dim % self.num_att_heads == 0, "input_dim must be divisible by num_att_heads"
+        assert self.dropout_broadcast_axes is None or self.dropout_broadcast_axes in [
+            "B",
+            "T",
+            "BT",
+        ], "invalid value, supported are None, 'B', 'T' and 'BT'"
+
+
+class ConformerMHSARelPosV1(nn.Module):
+    """
+    Conformer multi-headed self-attention module supporting
+        - relative positional encoding proposed by Shaw et al. (cf. https://arxiv.org/abs/1803.02155) by setting `learnable_pos_emb` to True and `with_pos_bias` to False
+        - and Transformer-XL style relative PE by Dai et al. (cf. https://arxiv.org/abs/1901.02860) by setting `learnable_pos_emb` to False and `with_pos_bias` to True
+
+    """
+
+    def __init__(self, cfg: ConformerMHSARelPosV1Config):
+
+        super().__init__()
+
+        self.layernorm = nn.LayerNorm(cfg.input_dim)
+
+        self.embed_dim = cfg.input_dim
+        self.num_heads = cfg.num_att_heads
+        self.embed_dim_per_head = self.embed_dim // self.num_heads
+
+        self.learnable_pos_emb = cfg.learnable_pos_emb
+        self.rel_pos_clip = cfg.rel_pos_clip
+        self.with_pos_bias = cfg.with_pos_bias
+        self.pos_emb_dropout = nn.Dropout(cfg.pos_emb_dropout)
+
+        assert not self.learnable_pos_emb or self.rel_pos_clip
+
+        self.att_weights_dropout = nn.Dropout(cfg.att_weights_dropout)
+
+        assert self.embed_dim % self.num_heads == 0, "embed_dim must be divisible by num_heads"
+
+        # projection matrices
+        self.q_proj_weight = nn.parameter.Parameter(torch.empty((self.embed_dim, self.embed_dim)))
+        self.k_proj_weight = nn.parameter.Parameter(torch.empty((self.embed_dim, self.embed_dim)))
+        self.v_proj_weight = nn.parameter.Parameter(torch.empty((self.embed_dim, self.embed_dim)))
+
+        self.in_proj_bias = nn.parameter.Parameter(torch.empty(3 * self.embed_dim))
```

Comment:
I don't understand, why do you keep the weight matrices for qkv separate, but then concat it for the bias? Also calling it `in_proj` was confusing to me, I would have called it `qkv_proj`. But anyway, why not simply use `nn.Linear`? So like this:
```suggestion
        self.qkv = nn.Linear(self.embed_dim, 3 * self.embed_dim)
```
Also see the RF code, which does it exactly like that.